qhbmlib/inference/ebm.py

```diff
@@ -520,3 +520,145 @@ def _log_partition_forward_pass(self):
   def _sample(self, num_samples: int):
     """See base class docstring"""
     return self.distribution.sample(num_samples, seed=self.seed)
+
+
+class GibbsWithGradientsKernel(tfp.mcmc.TransitionKernel):
+  """Implements the Gibbs With Gradients update rule.
+
+  See Algorithm 1 of
+  https://arxiv.org/abs/2102.04509v2
+  for details.
+  """
+
+  def __init__(self, input_energy):
+    """Initializes a GibbsWithGradientsKernel.
+
+    Args:
+      input_energy: The parameterized energy function which helps define the
+        acceptance probabilities of the Markov chain.
+    """
+    self._energy = input_energy
+    self._num_bits = input_energy.num_bits
+    self._parameters = dict(input_energy=input_energy)
+    self._q_i_of_x_probs = tf.Variable([0.0] * self._num_bits, trainable=False)
```

Comment:
I find that it might actually be clearer to use descriptive names rather than the notation of the Algorithm being implemented. One way to get the benefits of both would be as so:

```python
# q(i | x) in Algorithm 1
self.index_proposal_probs = tf.Variable([0.0] * self._num_bits, trainable=False)
```

same thought applies to other vars like d_tilde (e.g. `approx_energy_diff`) and so on

Reply:
I like the combination of comment with the math and more readable variable name, done